py3status/modules/year_progress.py

```diff
@@ -0,0 +1,212 @@
+    """Compute current year progress to next year"""
+    if not current:
+        current = datetime.now(tz=UTC())
+
+    start = datetime(current.year, 1, 1, tzinfo=UTC())
+    if not end:
+        end = datetime(current.year + 1, 1, 1, tzinfo=UTC())
+    return compute_progress(start, end, current)
+
+
+class Py3status:
+    format = "{progress_bar} {ratio}%{mode}"
+    progress_width = 5
+    progress_block = "▓"
+    remain_block = "░"
+    cache_timeout = 3600
+    progress_mode = "year"
+    ctime = None
+    thresholds = 90  # precentage, .e.g 90%
+
+    def _check_urgency(self, thresholds, progress):
+        urgency = False
+        print(progress)
+        if progress > thresholds:
+            urgency = True
+
+        return urgency
+
+    def _create_progress_string(self, progress, width=20):
+        progress_int = int(round(progress * width))
+        rest_int = int(width - progress_int)
+        return "{}{}".format(
+            self.progress_block * progress_int, self.remain_block * rest_int
+        )
+
+    def year_progress(self):
+        end_time = None
+        if self.ctime:
+            end_time = datetime.strptime(self.ctime[0], self.ctime[1]).astimezone()
+
+        if self.progress_mode == "year":
+            progress_ratio = compute_current_year_progress(end=end_time)
+            mode = "y"
+        elif self.progress_mode == "month":
+            progress_ratio = compute_current_month_progress(end=end_time)
+            mode = "m"
+        elif self.progress_mode == "day":
+            progress_ratio = compute_current_day_progress(end=end_time)
+            mode = "d"
+        elif self.progress_mode == "week":
+            progress_ratio = compute_current_week_progress()
+            if end_time:  # can't use custom end time
+                progress_ratio = None
+            mode = "w"
+
+        if progress_ratio:
+            ratio_int = int(progress_ratio * 100)
+            progress_bar = self._create_progress_string(
+                progress_ratio, width=self.progress_width
+            )
+            is_urgent = self._check_urgency(self.thresholds, ratio_int)
+        else:
+            progress_bar = ":)"
+            ratio_int = 0
+
+        color = None
+        if is_urgent:
+            color = self.py3.COLOR_BAD
+
+        data = {"progress_bar": progress_bar, "ratio": ratio_int, "mode": mode}
+        status = self.py3.safe_format(self.format, data)
+
+        return {
+            "full_text": status,
+            "color": color,
+            "cached_until": self.py3.time_in(self.cache_timeout),
+        }
+
+    def on_click(self, event):
+        if self.progress_mode == "year":
+            self.progress_mode = "month"
+        elif self.progress_mode == "month":
+            self.progress_mode = "day"
+        elif self.progress_mode == "day":
+            self.progress_mode = "week"
+        elif self.progress_mode == "week":
+            self.progress_mode = "year"
```

Comment:
(Cool). You can use dict to map things too.
```python
self.progress_mode = {                                                                
    "year": "month",
    "month": "day",
    "day": "week",
    "week": "year",
}.get(self.progress_mode, self.progress_mode)
```

I think it would be better to make a list and scroll over the list instead tho... because of addable custom dates (yah, plural with list `ctime`), but let's not worry about that anyway.

Reply:
Wow, I learn a lot. thanks!